library.cpp

```diff
@@ -1,518 +1,403 @@
+        }
+        strTo.resize(n);
+        return strTo;
 #endif
-}
-
-// Convert an UTF8 string to a wide Unicode String
-std::wstring utf8_decode(std::string const& str)
-{
-    if (str.empty()) {
-        return {};
     }
+
+    // Convert an UTF8 string to a wide Unicode String
+    std::wstring utf8_decode(std::string const& str)
+    {
+        if (str.empty()) {
+            return {};
+        }
 #if _WIN32 || _WIN64
-    int size_needed = MultiByteToWideChar(CP_UTF8, 0, &str[0], (int)str.size(), NULL, 0);
-    std::wstring wstrTo(size_needed, 0);
-    MultiByteToWideChar(CP_UTF8, 0, &str[0], (int)str.size(), &wstrTo[0], size_needed);
-    return wstrTo;
+        int          size_needed = MultiByteToWideChar(CP_UTF8, 0, &str[0], (int)str.size(), NULL, 0);
+        std::wstring wstrTo(size_needed, 0);
+        MultiByteToWideChar(CP_UTF8, 0, &str[0], (int)str.size(), &wstrTo[0], size_needed);
+        return wstrTo;
 #else
-    std::mbstate_t state{};
+        std::mbstate_t state{};
 
-    char const* in     = str.c_str();
-    size_t size_needed = std::mbsrtowcs(nullptr, &in, 0, &state);
+        char const* in          = str.c_str();
+        size_t      size_needed = std::mbsrtowcs(nullptr, &in, 0, &state);
 
-    std::wstring wstrTo(1 + size_needed, 0);
-    size_needed = std::mbsrtowcs(wstrTo.data(), &in, wstrTo.size(), &state);
+        std::wstring wstrTo(1 + size_needed, 0);
+        size_t n = std::mbsrtowcs(&wstrTo[0], &in, wstrTo.size(), &state);
 
-    if (size_needed == -1ul) {
-        throw std::domain_error("mbsrtowcs");
-    }
-    wstrTo.resize(size_needed);
-    return wstrTo;
+        if (n == -1ul) {
+            throw std::domain_error("mbsrtowcs");
+        }
+        wstrTo.resize(size_needed);
+        return wstrTo;
 #endif
-}
-
-/// Print error related information in stderr
-/// \param ec instance that contains error related information
-/// \param what customize prefix in output
-void fail(beast::error_code ec, wchar_t const* what) {
-    std::wcerr << what << L": " << utf8_decode(ec.message()) << std::endl;
-}
-
-class session : public std::enable_shared_from_this<session> {
-    tcp::resolver resolver_;
-    websocket::stream<beast::tcp_stream> ws_;
-    beast::flat_buffer buffer_;
-    std::wstring host_;
-    // path part in url. For example: /v2/ws
-    std::wstring path_;
-    std::wstring text_;
-    bool is_first_write_;
-
-public:
-    /// Resolver and socket require an io_context
-    /// \param ioc
-    explicit
-    session(net::io_context &ioc)
-            : resolver_(net::make_strand(ioc)), ws_(net::make_strand(ioc)), is_first_write_(false) {
-    }
-    /// Send message to remote websocket server
-    /// \param data to be sent
-    void send_message(const std::wstring &data) {
-        if(EnableVerbose)
-            std::wcout << L"<WsDll-" ARCH_LABEL ">> Sending message: " << data << std::endl;
-
-      const std::string to_send = utf8_encode(data);
```

Comment:
this is a local variable, it doesn't live long enough for `async_write` to complete